src/sasl/cyrsasl_external.erl

```diff
@@ -66,73 +76,150 @@ mech_step(#state{creds = Creds}, User) ->
-    case check_auth_req(XmppAddrs, CommonName, User, Server) of
+    VerificationList = get_verification_list(Server),
+    case verification_loop(VerificationList, Creds) of
         {error, Error} ->
             {error, Error};
         {ok, Name} ->
             NewCreds = mongoose_credentials:extend(Creds, [{username, Name}]),
             ejabberd_auth:authorize(NewCreds)
     end.
 
-check_auth_req([], CommonName, <<"">>, Server) ->
-    case ejabberd_auth:get_opt(Server, cyrsasl_external, standard) of
-       use_common_name when is_binary(CommonName) ->
-            {ok, CommonName};
-        _ ->
-            {error, <<"not-authorized">>}
-    end;
-check_auth_req([OneXmppAddr], _, <<"">>, Server) ->
-    verify_server(OneXmppAddr, Server);
-check_auth_req(_, _,  <<"">>, _) ->
-    {error, <<"not-authorized">>};
-check_auth_req([], undefined,  User, Server) ->
-    verify_server(User, Server);
-check_auth_req([], CommonName,  User, Server) ->
-    CNOption = ejabberd_auth:get_opt(Server, cyrsasl_external, standard),
-    maybe_use_common_name(CommonName, User, Server, CNOption);
-check_auth_req([_], _,  _, _) ->
-    {error, <<"invalid-authzid">>};
-check_auth_req(XmppAddrs, _,  User, Server) ->
-    case lists:member(User, XmppAddrs) of
-        true ->
-            verify_server(User, Server);
-        _ ->
-            {error, <<"not-authorized">>}
+get_verification_list(Server) ->
+    case ejabberd_auth:get_opt(Server, cyrsasl_external, [standard]) of
+        [] -> [standard];
+        List when is_list(List) -> List;
+        standard -> [standard];
+        use_common_name -> [standard, common_name];
+        allow_just_user_identity -> [standard, auth_id]
+    end.
+
+verification_loop([VerificationFn | T], Creds) ->
+    case verify_creds(VerificationFn, Creds) of
+        {error, Error} when T =:= [] ->
+            {error, Error};
+        {error, _} ->
+            verification_loop(T, Creds);
+        {ok, Name} ->
+            {ok, Name}
     end.
 
-maybe_use_common_name(_, User, Server, allow_just_user_identity) ->
-    verify_server(User, Server);
-maybe_use_common_name(CommonName, User, Server, use_common_name) ->
-    case verify_server(User, Server) of
-        {ok, CommonName} ->
+verify_creds(standard, Creds) ->
+    standard_verification(Creds);
+verify_creds(common_name, Creds) ->
+    common_name_verification(Creds);
+verify_creds(auth_id, Creds) ->
+    auth_id_verification(Creds);
+verify_creds({mod, Mod}, Creds) ->
+    custom_verification(Mod, Creds).
+
+
+standard_verification(Creds) ->
+    Server = mongoose_credentials:lserver(Creds),
+    XmppAddrs = get_credentials(Creds, xmpp_addresses),
+    AuthId = get_credentials(Creds, auth_id),
+    XmppAddr = case {XmppAddrs, AuthId} of
+                   {[OneXmppAddr], undefined} ->
+                       OneXmppAddr;
+                   {[_], _} ->
+                       {error, <<"invalid-authzid">>};
+                   {[_, _ | _], undefined} ->
```

Comment:
You can merge those two cases to one, right?
```erlang
{[_], _} ->
    {error, <<"invalid-authzid">>};
{[_, _ | _], undefined} ->
    {error, <<"invalid-authzid">>};
```
```erlang
{[ _ | _ ], _} ->
    {error, <<"invalid-authzid">>};
```

Reply:
no, these 2 are different:
- is first one is when only one address is specified in `XmppAddrs` but some address is also provided in `AuthId`
- the second one is when more than one address is specified in `XmppAddrs` but nothing is provided in `AuthId`

-

Reply:
thanks for explanation :)